### PR TITLE
Improve docs of bit twiddling functions

### LINKED ISSUE #394: Bit twiddling functions should have better documentation
I've been struggling to understand them for a while. Some examples should help.

### PR DESCRIPTION
**Pull request by sjakobi**
Fixes #394.

---

TODO:

* [x] Add docs for `Hash`, `Bitmap`, `Shift` types
* [x] ~~Add `incrementShift s = s + bitsPerSubkey`?!~~ Moved to #426.

### CODE REVIEW

File: Data/HashMap/Internal.hs [hunk 1]
```diff
@@ -2247,34 +2247,62 @@ clone ary =
...
+-- TODO: Name this 'bitsPerLevel'?! What's a "subkey"?
 bitsPerSubkey :: Int
 bitsPerSubkey = 5
 
+-- | The size of a 'Full' node, i.e. @2 ^ 'bitsPerSubkey'@.
 maxChildren :: Int
 maxChildren = 1 `unsafeShiftL` bitsPerSubkey
 
+-- | Bitmap with the lowest 'bitsPerSubkey' bits set, i.e. @0b11111@.
 subkeyMask :: Bitmap
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:2259]: We never use this mask as a bitmap, so it should probably simply be a `Word` or maybe a new `Mask` type.

Comment (treeowl) [reply Data/HashMap/Internal.hs:2259]: Agreed.

Comment (sjakobi) [reply Data/HashMap/Internal.hs:2259]: I realized that `subkeyMask` is used only once: in `index`. It's probably not worth adding a new type for it, so I'll go with `Word`.

**Comment by sjakobi** [comment]
I'll leave any refactorings for future work.

**Pull request merged**